### PR TITLE
Add support for query params with ref schema in service generation

### PR DESCRIPTION
**Pull request by aneeshafedo**
Resolves https://github.com/ballerina-platform/openapi-tools/issues/1331

## Goals

Example:

```yaml
paths:
  /users/meetings:
    get:
      description: List all the meetings that were scheduled
      operationId: listMeetings
      parameters:
        - description: "Meeting location"
          in: query
          name: location
          schema:
            $ref: '#/components/schemas/RoomNo'
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MeetingList'
          description: "HTTP Status Code:200. List of meetings returned."
components:
  schemas:
    RoomNo:
      enum:
        - R1
        - R3
        - R5
      type: string
      nullable: true
```

**Proposed generation**

```bal
public type RoomNo string;

     resource function get users/meetings(RoomNo location) {
     }

```

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java [hunk 1]
```diff
@@ -81,8 +85,14 @@ public class ParametersGenerator {
     private final List<Node> requiredParams;
     private final List<Node> defaultableParams;
 
-    public ParametersGenerator(boolean isNullableRequired) {
+    private final OpenAPI openAPI;
+
+    private static final List<String> queryParamSupportedTypes =
+            new ArrayList<>(Arrays.asList(GeneratorConstants.INTEGER, GeneratorConstants.NUMBER,
+                    GeneratorConstants.STRING, GeneratorConstants.BOOLEAN));
+    public ParametersGenerator(boolean isNullableRequired, OpenAPI openAPI) {
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:93]: Missing newline

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ServiceDiagnosticMessages.java [hunk 2]
```diff
@@ -47,7 +47,9 @@ public enum ServiceDiagnosticMessages {
             "Header '%s' with no header type can not be mapped to the Ballerina headers.",
             DiagnosticSeverity.ERROR),
     OAS_SERVICE_107("OAS_SERVICE_107", "HTTP status code '%s' is not supported in Ballerina.",
-            DiagnosticSeverity.ERROR);
+            DiagnosticSeverity.ERROR),
+    OAS_SERVICE_108("OAS_SERVICE_108", "Only array types of string, int, float, boolean, " +
+            "and decimal are allowed to be used in query parameters in Ballerina.", DiagnosticSeverity.ERROR);
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ServiceDiagnosticMessages.java:52]:
```suggestion
            "and decimal are allowed to be used as query parameters in Ballerina.", DiagnosticSeverity.ERROR);
```

**Comment by NipunaRanasinghe** [comment]
@aneeshafedo lets attach the related issue to the PR description

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java (hunk 1 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java]: Resolved in https://github.com/ballerina-platform/openapi-tools/pull/1335/commits/445e23f5bd95629f18e244673178a695e7d65eda

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ServiceDiagnosticMessages.java (hunk 2 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ServiceDiagnosticMessages.java]: Resolved in https://github.com/ballerina-platform/openapi-tools/pull/1335/commits/445e23f5bd95629f18e244673178a695e7d65eda

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java [hunk 3]
```diff
@@ -364,12 +398,30 @@ private Node handleOptionalQueryParameter(Schema<?> schema, NodeList<AnnotationN
...
+                    createToken(SyntaxKind.EQUAL_TOKEN),
+                    createSimpleNameReferenceNode(createIdentifierToken(defaultValue)));
+        } else if (parameter.getRequired() && (refSchema.getNullable() == null || (!refSchema.getNullable()))) {
+            return createRequiredParameterNode(annotations, rTypeName, parameterName);
+        } else {
+            OptionalTypeDescriptorNode optionalNode = createOptionalTypeDescriptorNode(rTypeName,
+                    createToken(SyntaxKind.QUESTION_MARK_TOKEN));
+            return createRequiredParameterNode(annotations, optionalNode, parameterName);
+        }
+    }
```
Comment (lnash94) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:418]: Missing new line

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java [hunk 4]
```diff
@@ -271,4 +272,49 @@ public void optionalQueryParameter() throws IOException, BallerinaOpenApiExcepti
...
+                .withFilters(filter)
+                .build();
+        BallerinaServiceGenerator ballerinaServiceGenerator = new BallerinaServiceGenerator(oasServiceMetadata);
+        syntaxTree = ballerinaServiceGenerator.generateSyntaxTree();
+    }
+
+    @Test(description = "18. Query parameter(s) having a referenced schema of array of unsupported type",
+            expectedExceptions = BallerinaOpenApiException.class,
+            expectedExceptionsMessageRegExp = "Only array types of string, int, float, boolean, and decimal are " +
+                    "allowed to be used as query parameters in Ballerina.")
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java:309]: Here, can't we unify these 2 similar error messages to follow a common pattern? 

how about something like `Type '<NAME>' is not a valid query parameter type in Ballerina. Only <LIST_OF_TYPES> are supported.`?

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java [hunk 5]
```diff
@@ -364,12 +398,30 @@ private Node handleOptionalQueryParameter(Schema<?> schema, NodeList<AnnotationN
         }
     }
 
+    private Node handleReferencedQueryParameter(Parameter parameter, String refTypeName, Schema<?> refSchema,
+                                                NodeList<AnnotationNode> annotations, IdentifierToken parameterName) {
+        BuiltinSimpleNameReferenceNode rTypeName = createBuiltinSimpleNameReferenceNode(null,
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:403]:
```suggestion
        BuiltinSimpleNameReferenceNode refTypeName = createBuiltinSimpleNameReferenceNode(null,
```

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java (hunk 3 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:418]: Updated in https://github.com/ballerina-platform/openapi-tools/pull/1335/commits/227a2678799ce9dbf3cdad98015df7107add3b4d

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java (hunk 4 above)
Comment (aneeshafedo) [reply openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java]: Updated in https://github.com/ballerina-platform/openapi-tools/pull/1335/commits/227a2678799ce9dbf3cdad98015df7107add3b4d

Message updated as below

```
Type '<TypeName>' is not a valid query parameter type in Ballerina. The supported types are string, int, float, boolean, decimal, array types of the aforementioned types and map<json>.
```

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java (hunk 5 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java]: Updated https://github.com/ballerina-platform/openapi-tools/commit/227a2678799ce9dbf3cdad98015df7107add3b4d

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply openapi-cli/src/test/java/io/ballerina/openapi/generators/service/QueryParameterTests.java]: +1

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Review by lnash94 — APPROVED** [review]

**Pull request merged**